--- a/kirki-packages/data-option/src/Option.php
+++ b/kirki-packages/data-option/src/Option.php
@@ -89,7 +89,8 @@
 				 * We need to change the value accordingly depending on whether
 				 * this is the last item in the loop or not.
 				 */
-				$value = ( isset( $parts[ $key + 1 ] ) ) ? [] : '';
+				// $value = ( isset( $parts[ $key + 1 ] ) ) ? [] : '';
+				$value = ( isset( $parts[ $key + 1 ] ) ) ? [] : $default;
 			}
 		}
 
