// gulp-stylelint/dist/index.js
'use strict';

var _stylelint = require('stylelint');

var _pluginError = require('plugin-error');

var _pluginError2 = _interopRequireDefault(_pluginError);

var _through = require('through2');

var _through2 = _interopRequireDefault(_through);

var _promise = require('promise');

var _promise2 = _interopRequireDefault(_promise);

var _deepExtend = require('deep-extend');

var _deepExtend2 = _interopRequireDefault(_deepExtend);

var _reporterFactory = require('./reporter-factory');

var _reporterFactory2 = _interopRequireDefault(_reporterFactory);

var _applySourcemap = require('./apply-sourcemap');

var _applySourcemap2 = _interopRequireDefault(_applySourcemap);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Name of this plugin for reporting purposes.
 * @type {String}
 */
var pluginName = 'gulp-stylelint';

/**
 * Stylelint results processor.
 * @param {Object} [options] - Plugin options.
 * @param {String} [options.reportOutputDir] - Common path for all reporters.
 * @param {[Object]} [options.reporters] - Reporter configurations.
 * @param {Boolean} [options.debug] - If true, error stack will be printed.
 * @return {Stream} Object stream usable in Gulp pipes.
 */
/**
 * Gulp stylelint plugin.
 * @module gulp-stylelint
 */

module.exports = function gulpStylelint(options) {
  // eslint-disable-line max-statements

  /**
   * Plugin options with defaults applied.
   * @type Object
   */
  var pluginOptions = (0, _deepExtend2.default)({
    failAfterError: true,
    debug: false
  }, options);

  /**
   * Lint options for stylelint's `lint` function.
   * @type Object
   */
  var lintOptions = (0, _deepExtend2.default)({}, options);

  /**
   * List of gulp-stylelint reporters.
   * @type [Function]
   */
  var reporters = (pluginOptions.reporters || []).map(function (config) {
    return (0, _reporterFactory2.default)(config, pluginOptions);
  });

  /**
   * List of stylelint's lint result promises.
   * @type [Promise]
   */
  var lintPromiseList = [];

  // Remove the stylelint options that cannot be used:
  delete lintOptions.files; // css code will be provided by gulp instead
  delete lintOptions.formatter; // formatters are defined in the `reporters` option
  delete lintOptions.cache; // gulp caching should be used instead

  // Remove gulp-stylelint options so that they don't interfere with stylelint options:
  delete lintOptions.reportOutputDir;
  delete lintOptions.reporters;
  delete lintOptions.debug;

  /**
   * Launches linting of a given file, pushes promises to the promise list.
   *
   * Note that the files are not modified and are pushed
   * back to their pipes to allow usage of other plugins.
   *
   * @param {File} file - Piped file.
   * @param {String} encoding - File encoding.
   * @param {Function} done - File pipe completion callback.
   * @return {undefined} Nothing is returned (done callback is used instead).
   */
  function onFile(file, encoding, done) {

    if (file.isNull()) {
      done(null, file);
      return;
    }

    if (file.isStream()) {
      this.emit('error', new _pluginError2.default(pluginName, 'Streaming is not supported'));
      done();
      return;
    }

    var localLintOptions = (0, _deepExtend2.default)({}, lintOptions, {
      code: file.contents.toString(),
      codeFilename: file.path
    });

    var lintPromise = (0, _stylelint.lint)(localLintOptions).then(function (lintResult) {
      return (
        // Checking for the presence of sourceMap.mappings
        // in case sourcemaps are initialized, but still empty:
        file.sourceMap && file.sourceMap.mappings ? (0, _applySourcemap2.default)(lintResult, file.sourceMap) : lintResult
      );
    }).then(function (lintResult) {
      if (lintOptions.fix && lintResult.output) {
        file.contents = Buffer.from(lintResult.output);
      }
      done(null, file);
      return lintResult;
    }).catch(function (error) {
      done(null, file);
      return _promise2.default.reject(error);
    });

    lintPromiseList.push(lintPromise);
  }

  /**
   * Provides Stylelint result to reporters.
   * @param {[Object]} lintResults - Stylelint results.
   * @return {Promise} Resolved with original lint results.
   */
  function passLintResultsThroughReporters(lintResults) {
    var warnings = lintResults.reduce(function (accumulated, res) {
      return accumulated.concat(res.results);
    }, []);
    return _promise2.default.all(reporters.map(function (reporter) {
      return reporter(warnings);
    })).then(function () {
      return lintResults;
    });
  }

  /**
   * Determines if the severity of a stylelint warning is "error".
   * @param {Object} warning - Stylelint results warning.
   * @return {Boolean} True if warning's severity is "error", false otherwise.
   */
  function isErrorSeverity(warning) {
    return warning.severity === 'error';
  }

  /**
   * Resolves promises and provides accumulated report to reporters.
   * @param {Function} done - Stream completion callback.
   * @return {undefined} Nothing is returned (done callback is used instead).
   */
  function onStreamEnd(done) {
    var _this = this;

    _promise2.default.all(lintPromiseList).then(passLintResultsThroughReporters).then(function (lintResults) {
      process.nextTick(function () {
        var errorCount = lintResults.reduce(function (sum, res) {
          var errors = res.results[0].warnings.filter(isErrorSeverity);
          return sum + errors.length;
        }, 0);
        if (pluginOptions.failAfterError && errorCount > 0) {
          var errorMessage = 'Failed with ' + errorCount + ' ' + (errorCount === 1 ? 'error' : 'errors');
          _this.emit('error', new _pluginError2.default(pluginName, errorMessage));
        }
        done();
      });
    }).catch(function (error) {
      process.nextTick(function () {
        _this.emit('error', new _pluginError2.default(pluginName, error, {
          showStack: Boolean(pluginOptions.debug)
        }));
        done();
      });
    });
  }

  return _through2.default.obj(onFile, onStreamEnd).resume();
};

/**
 * Formatters bundled with stylelint by default.
 *
 * User may want to see the list of available formatters,
 * proxy them or pass them as functions instead of strings.
 *
 * @see https://github.com/olegskl/gulp-stylelint/issues/3#issuecomment-197025044
 * @type {Object}
 */
module.exports.formatters = _stylelint.formatters;
